### PR TITLE
consolidate and improve internal rest infrastructure

### PR DESCRIPTION
**Pull request by Plerx2493**
# Summary

Rewoks the whole rest client to be faster and slightly more accurate

# Details

- switches from internally using Task to ValueTask
- completely reworks the entire request handling model
- respects which endpoints are exempt from the global rate limit
- annotates the entire API client for nullability

## Future Notes

this PR does NOT update the rest client to match the current API specification yet, that work will have to be done in a later PR

### CODE REVIEW

File: DSharpPlus/Net/Rest/RestClient.cs [hunk 1]
```diff
@@ -22,692 +22,203 @@
...
-                this.UpdateHashCaches(request, bucket);
-                bucket.Maximum = 0;
-                bucket._remaining = 0;
-                return;
-            }
-
-            // no need to wait on all the potentially waiting tasks
-            _ = Task.Run(() => ratelimitTcs.TrySetResult(false));
-        }
+        // retrying forever is rather suboptimal, but it's the old behaviour. We should discuss whether
```
Comment (Plerx2493) [inline DSharpPlus/Net/Rest/RestClient.cs:101]: Maybe we can add a switch to the discordconfig

**Comment by akiraveliara** [comment]
should be ready to review, cc @VelvetToroyashi; with how large this PR is there *might* be a case for rebase-merging so we can revert individual commits, but i'm not sure whether that makes too much sense here

**Review by Naamloos — APPROVED** [review]

File: Directory.Packages.props [hunk 2]
```diff
@@ -1,14 +1,13 @@
 <Project>
   <ItemGroup>
-    <PackageVersion Include="ConcurrentHashSet" Version="1.3.0"/>
-    <PackageVersion Include="DSharpPlus.VoiceNext.Natives" Version="1.0.0"/>
-    <PackageVersion Include="Microsoft.CodeAnalysis.CSharp.Scripting" Version="4.6.0-1.final"/>
-    <PackageVersion Include="Microsoft.Extensions.Caching.Memory" Version="7.0.0"/>
-    <PackageVersion Include="Microsoft.Extensions.DependencyInjection" Version="7.0.0"/>
-    <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="7.0.0"/>
-    <PackageVersion Include="Microsoft.Extensions.Logging.Abstractions" Version="7.0.1"/>
-    <PackageVersion Include="Microsoft.SourceLink.GitHub" Version="1.1.1" PrivateAssets="all"/>
```
Comment (InFTord) [inline Directory.Packages.props:10]: re-add SourceLink, required for our docs bot to function iirc

**Review by InFTord — APPROVED** [review]
LGTM, besides one small nitpick

**Pull request merged**